docs/python/tutorial-fastapi.md

```diff
@@ -0,0 +1,652 @@
+
+We'll now define routes in our FastAPI application. In the context of web applications, routes are like pathways that map specific URLs to the code that handles them. These routes serve as the entry points for different functionalities within our application. When a client, such as a web browser or another program, sends a request to our application with a particular URL, FastAPI routes that request to the appropriate function (also known as route handler or view function) based on the URL, and that function processes the request and generates a response.
+
+Let's proceed with defining routes that will allow us to add and retrieve individual items, as well as list all items in the grocery list.
+
+3. Add the following route at the end of the `main.py` file:
+    ```python
+    # Route to add a item
+    @app.post("/items/{item_name}/{quantity}")
+    def add_item(item_name: str, quantity: int):
+        if quantity <= 0:
+            raise HTTPException(status_code=400, detail="Quantity must be greater than 0.")
+        # if item already exists, we'll just add the quantity.
+        # get all item names
+        items_ids = {item.item_name: item.item_id if item.item_id is not None else 0 for item in grocery_list.values()}
+        if item_name in items_ids.keys():
+            # get index of item_name in item_ids, which is the item_id
+            item_id = items_ids[item_name]
+            grocery_list[item_id].quantity += quantity
+    # otherwise, create a new item
+        else:
+            # generate an id for the item based on the highest ID in the grocery_list
+            item_id = max(grocery_list.keys()) + 1 if grocery_list else 0
+            grocery_list[item_id] = ItemPayload(
+                item_id=item_id, item_name=item_name, quantity=quantity
+            )
+
+        return {"item": grocery_list[item_id]}
+    ```
+
+    If you have enabled type hints, you might notice Pylance adds inlay hints with the function return type, as well as the types for `item_ids` and `item_id`. You can optionally double click on each suggestion to insert them into the code:
+    ![Inlay function return and variable type hints being displayed by Pylance throughout the sample code.](images/fastapi-tutorial/pylance_inlay_hints.png)
+
+Now let's see if this route is working as expected. The fastest way to do so is to leverage both VS Code's debugger as well as FastAPI's `/docs` endpoint, which provides information about all the available API routes and allows you to interact with the API to explore their parameters and responses. This documentation is generated dynamically based on the metadata and type hints defined in the FastAPI application.
+
+4. Add a breakpoint next to the if quantity <= 0 statement, by clicking on the left margin of the line number (or `kb(editor.debug.action.toggleBreakpoint)`). This will make the debugger stop prior to the execution of that line, allowing us to inspect the code line by line.
```

Comment:
```suggestion
4. Add a breakpoint next to the if `quantity <= 0` statement, by clicking on the left margin of the line number (or `kb(editor.debug.action.toggleBreakpoint)`). This will make the debugger stop prior to the execution of that line, allowing us to inspect the code line by line.
```